src/cartservice/Program.cs

```diff
@@ -163,21 +143,48 @@ static void Main(string[] args)
             }
         }
 
-        private static string ReadRedisAddress(string address)
+        /// <summary>
+        /// Reads parameter in the right order
+        /// </summary>
+        /// <param name="description">Parameter description</param>
+        /// <param name="commandLineValue">Value provided from the command line</param>
+        /// <param name="environmentVariableName">The name of environment variable where it could have been set</param>
+        /// <param name="environmentParser">The method that parses environment variable and returns typed parameter value</param>
+        /// <param name="defaultValue">Parameter's default value - in case other method failed to assign a value</param>
+        /// <typeparam name="T">The type of the parameter</typeparam>
+        /// <returns>Parameter value read from all the sources in the right order(priorities)</returns>
+        private static T ReadParameter<T>(
+            string description,
+            T commandLineValue, 
+            string environmentVariableName, 
+            Func<string, T> environmentParser, 
+            T defaultValue)
         {
-            if (!string.IsNullOrEmpty(address))
+            // Command line argument
+            if(!EqualityComparer<T>.Default.Equals(commandLineValue, default(T))) 
             {
-                return address;
+                return commandLineValue;
             }
 
-            Console.WriteLine($"Reading redis cache address from environment variable {REDIS_ADDRESS}");
-            string redis = Environment.GetEnvironmentVariable(REDIS_ADDRESS);
-            if (!string.IsNullOrEmpty(redis))
+            // Environment variable
+            Console.Write($"Reading {description} from environment variable {environmentVariableName}. ");
```

Comment:
Perhaps use ILogger instead of Console here?

Reply:
Do you mean ILogger from Microsoft.Extensions.Logging nuget? 
In this sample we don't use ASP.NET - cart microservice uses grpc (and grpc host) for communication and overall design tries to follow 12 factor apps design, where logs should go to std.out, so sidecars can read them and transparently transfer them out. So we don't use Microsoft.Extensions.* packages or any other dependencies that are not absolutely required.

Reply:
`ILogger` is easy configurable for console or debug output. Moreover many libraries already use ILogger. So I'd encourage to use it here.

``` csharp
serviceCollection.AddSingleton(new LoggerFactory()
    .AddConsole()
    .AddDebug());
```

Reply:
This is true for ASP.NET Core with built-in dependency injection. Grpc C# library doesn't support dependency injection and so we don't have it here too.